scenes/cslib_object.txt

```diff
@@ -0,0 +1,111 @@
+*bug cslib_object is a cslib module and must not be accessed directly
+
+*comment cslib_object
+*comment Allows the simple manipulation of arrays of sets of variables
+*comment
+*comment How it works:
+*comment A object is always accessed through its main instance.
+*comment When you "get" an instance of a object, the data for that particular instance 
+*comment is copied into the main instance
+*comment
+*comment A object definition has the following elements defined in startup.txt:
+*comment	1. A object "fieldname" array defining all the field names
+*comment		objectname_fieldname_N fieldname
+*comment	2. A main instance of this object. This is used to 'return'
+*comment		the object (similar to cslib_ret).
+*comment	3. An array of instances of the object
+*comment		
+*comment Public methods
+*comment	get object index
+*comment		copy the instance of the object at index into main instance
+*comment	set object index field value
```

Comment:
It would be nice IMO to update the whole object in one operation. After you manipulate the local dragon (e.g HP diminishes after an attack), you could simply copy back the whole thing, from the local instance back to the array.

Reply:
That's an interesting idea... Do you mean like so?

```
*gosub_scene cslib_object get "dragon" 1
${dragon_hp} = 200
*set dragon_hp (dragon_hp  - 100)
*gosub_scene cslib_object update "dragon" 1
${dragon_1_hp} = 100
```

I like the general idea of an update function, but I wonder if the pattern of editing the main instance will be confusing to the average CS author?

Reply:
Not sure about this. You would be locking the user into changing the base instance, which isn't necessarily what you want.

For example,  you fight a dragon and reduce it to 20 hit points. It flies away injured to recover. When you meet it again two weeks later, it has its original hit points back. If we have changed it, we no longer have any knowledge of what it originally was.

We could get round this by doubling up the data so that we have 'fixed' and 'volatile' versions of the data but this adds a lot of overhead and the user would then have to tell us whether it should be a volatile or a fixed update - another level of complication that could go wrong.

My feeling is that we should let the user decide on whether the data is save. Either they explicitly 'set' it, or the data doesn't change.

Reply:
I think if we implement the "Set Pairs" functionality, then there is no longer any need for this.
I'm largely in agreement with you though: I think the get/set pattern will be easier for CS authors to follow over the update pattern.